miniconf_cli/src/lib.rs

```diff
@@ -0,0 +1,309 @@
+                    debug_assert!(!node.is_leaf());
+                    ("...\n".as_bytes(), &mut buf[..])
+                }
+                Err(miniconf::Error::Traversal(Traversal::Absent(_))) => {
+                    continue;
+                }
+                ret => {
+                    debug_assert!(node.is_leaf());
+                    let (val, rest) = buf.split_at_mut(ret?) as _;
+                    (val as _, rest)
+                }
+            };
+            let check: u32 = yafnv::fnv1a(val.iter().copied());
+            awrite(&mut write, "  ".as_bytes()).await?;
+            let rl = rest.len();
+            let mut sl = &mut rest[..];
+            Path::<_, SEPARATOR>::from(WriteWrap(&mut sl)).transcode::<M, Y, _>(keys)?;
+            let path_len = rl - sl.len();
+            awrite(&mut write, &rest[root_len..path_len]).await?;
+            awrite(&mut write, ": ".as_bytes()).await?;
+            awrite(&mut write, val).await?;
+            let def = match def.get_json_by_key(keys, &mut buf[..]) {
+                Err(miniconf::Error::Traversal(Traversal::TooShort(_depth))) => {
+                    debug_assert!(!node.is_leaf());
+                    continue;
+                }
+                Err(miniconf::Error::Traversal(Traversal::Absent(_depth))) => "absent".as_bytes(),
+                ret => &buf[..ret?],
+            };
+            if yafnv::fnv1a::<u32, _>(def.iter().copied()) == check {
+                awrite(&mut write, " (default)\n".as_bytes()).await?;
+            } else {
+                awrite(&mut write, " (default: ".as_bytes()).await?;
+                awrite(&mut write, def).await?;
+                awrite(&mut write, ")\n".as_bytes()).await?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
+    use miniconf::Tree;
+
+    use super::*;
+
+    #[derive(Tree, Default)]
+    struct Inner {
+        e: i32,
+    }
+
+    #[derive(Tree, Default)]
+    struct Set {
+        a: i32,
+        #[tree(depth = 1)]
+        b: [i32; 3],
+        #[tree(depth = 1)]
+        c: Option<i32>,
+        #[tree(depth = 1)]
+        d: Inner,
+        #[tree(depth = 2)]
+        f: [Inner; 2],
+        #[tree(depth = 4)]
+        g: [[[Inner; 1]; 1]; 1],
+    }
+    const Y: usize = 5;
+    const S: usize = 128;
+
+    #[test]
+    fn new() {
+        let m = Menu::<Set, Y>::default();
+        for p in m.list::<S>().unwrap() {
+            println!("{}", p.unwrap());
+        }
+    }
+
+    #[tokio::test]
+    async fn all() {
+        let mut buf = [0; 1024];
+        let mut s = Set::default();
+        let mut stdout = embedded_io_adapters::tokio_1::FromTokio::new(tokio::io::stdout());
+        let mut menu = Menu::<Set, Y>::default();
+        s.c = Some(8);
+        menu.enter("b/0").unwrap();
+        menu.set(&mut s, b"1234").unwrap();
+        menu.exit(2).unwrap();
+        menu.push("f/1/e").unwrap().0.set(&mut s, b"9").unwrap();
```

Comment:
Same, shouldn't we check the original struct and verify it was set properly?

Reply:
These were more playground to see how it'll look that actual unittests. Hopefully they'll improve .. ;)